hoglet.go

```diff
@@ -136,21 +136,27 @@ func (c *Circuit[IN, OUT]) stateForCall() State {
 		// We reset openedAt to block further calls to pass through when half-open. A success will cause the breaker to
 		// close. This is slightly racy: multiple goroutines may reach this point concurrently since we do not lock the
 		// breaker.
-		c.setOpenedAt(time.Now().UnixMicro())
+		c.reopen()
 	}
 
 	return state
 }
 
-// setOpenedAt sets the time the circuit was opened at.
-// Passing a value of 0 closes the cirtuit.
-func (c *Circuit[IN, OUT]) setOpenedAt(i int64) {
-	if i == 0 {
-		c.openedAt.Store(0)
-	} else {
-		// CompareAndSwap is needed to avoid clobbering another goroutine's openedAt value.
-		c.openedAt.CompareAndSwap(0, i)
-	}
+// open marks the circuit as open, if it not already.
+// It is safe for concurrent calls and only the first one will actually set opening time.
+func (c *Circuit[IN, OUT]) open() {
+	// CompareAndSwap is needed to avoid clobbering another goroutine's openedAt value.
+	c.openedAt.CompareAndSwap(0, time.Now().UnixMicro())
+}
+
+// reopen forcefully (re)marks the circuit as open, resetting the half-open time.
+func (c *Circuit[IN, OUT]) reopen() {
```

Comment:
Do we need this? Why not just use `open`?

Reply:
because `open` does `CompareAndSwap`, so it won't "move" the opening time of a circuit when multiple calls run concurrently and one happens to take longer than the others (this would in turn cause the half-open delay to take longer). This is what we normally want.

But we need `reopen` for the case where we actually _want_ to move the opening time: when we just hit the half-open time.

Reply:
That was the whole reason for the bug we had: we were never resetting the opening time, so we never left the "half-open" state after it first triggered.